src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java

```diff
@@ -5013,14 +5018,69 @@ protected JCTree methodDeclaratorRest(int pos,
                 if (token.kind == DEFAULT) {
                     accept(DEFAULT);
                     defaultValue = annotationValue();
+                    accept(SEMI);
                 } else {
                     defaultValue = null;
+                    accept(SEMI, tk -> Errors.Expected2(LBRACE, SEMI));
                 }
-                accept(SEMI);
                 if (token.pos <= endPosTable.errorEndPos) {
                     // error recovery
-                    skip(false, true, false, false);
+                    // look if there is a probable missing opening brace,
+                    // and if yes, parse as a block
+                    skip(false, true, !unclosedParameterList, !unclosedParameterList);
+                    boolean parseAsBlock;
                     if (token.kind == LBRACE) {
+                        parseAsBlock = true;
+                    } else if (unclosedParameterList) {
+                        parseAsBlock = false;
+                    } else {
+                        parseAsBlock = switch (token.kind) {
+                            //definitelly sees a statement:
+                            case CASE, DEFAULT, IF, FOR, WHILE, DO, TRY, SWITCH,
+                                 RETURN, THROW, BREAK, CONTINUE, ELSE, FINALLY,
+                                 CATCH, THIS, SUPER, NEW -> true;
+                            case RBRACE -> {
+                                //check if adding an opening brace would balance out
+                                //the opening and closing braces:
+                                int braceBalance = 1;
+                                VirtualScanner virtualScanner = new VirtualScanner(S);
```

Comment:
This seems a new pattern - in other case we use the lookahead method. I don't object to this, as much as note that we now seem to use different styles to lookahead.